### ISSUE TITLE
Clearly document how to re-use allocations

### ISSUE THREAD

**Issue by mickvangelderen** [capnproto/capnproto#2181]
Allocation re-use was [discussed here](https://github.com/capnproto/capnproto/issues/639) and the suggested solution is to pass some memory into the `MallocMessageBuilder` constructor. 

I would love it if the following was documented:

1. Can you reset the arena inside a `MallocMessageBuilder` so it can be re-used?
2. Can you re-use memory by re-initializing all of the fields in a message root obtained with `MessageBuilder::initRoot`?

**Comment by kentonv** [comment]
The arena allocation nature of message builders is discussed in the docs here:

https://capnproto.org/cxx.html#tips-and-best-practices

The API reference documentation is the header comments.

If you feel either of these are missing important details (entirely possible!), it would be helpful if you could make suggestions in the form of a PR, as that makes it easier for me to understand what's missing. It can be difficult for me, as the author of the code, to form a mental model of what's obvious and what isn't.

FWIW this API hasn't changed in >10 years so the discussion you linked is still correct.

**Comment by mickvangelderen** [comment]
> Messages are built in “arena” or “region” style: each object is allocated sequentially in memory, until there is no more room in the segment, in which case a new segment is allocated, and objects continue to be allocated sequentially in that segment. This design is what makes Cap’n Proto possible at all, and it is very fast compared to other allocation strategies. However, it has the disadvantage that if you allocate an object and then discard it, that memory is lost. In fact, the empty space will still become part of the serialized message, even though it is unreachable. The implementation will try to zero it out, so at least it should pack well, but it’s still better to avoid this situation. Some ways that this can happen include:
> 
> - If you init a field that is already initialized, the previous value is discarded.
> - If you create an orphan that is never adopted into the message tree.
> - If you use adoptWithCaveats to adopt an orphaned struct into a struct list, then a shallow copy is necessary, since the struct list requires that its elements are sequential in memory. The previous copy of the struct is discarded (although child objects are transferred properly).
> - If you copy a struct value from another message using a set method, the copy will have the same size as the original. However, the original could have been built with an older version of the protocol which lacked some fields compared to the version your program was built with. If you subsequently get that struct, the implementation will be forced to allocate a new (shallow) copy which is large enough to hold all known fields, and the old copy will be discarded. Child objects will be transferred over without being copied – though they might suffer from the same problem if you get them later on. Sometimes, avoiding these problems is too inconvenient. Fortunately, it’s also possible to clean up the mess after-the-fact: > - if you copy the whole message tree into a fresh MessageBuilder, only the reachable objects will be copied, leaving out all of the unreachable dead space.
> 
> In the future, Cap’n Proto may be improved such that it can re-use dead space in a message. However, this will only improve things, not fix them entirely: fragmentation could still leave dead space.

From this discription it seems like calling `initFoo()` multiple times may cause a new piece of the arena to be allocated every time. That means that you can not reuse a `MessageBuilder`?

**Comment by kentonv** [comment]
Correct, you have to destroy the MessageBuilder and create a new one (potentially with the same scratch buffer). This is cheap to do -- no memory allocation is involved, as long as your messages never exceed the scratch space.
